Dependencies/xr/Source/ARKit/XR.mm

```diff
@@ -0,0 +1,462 @@
+#include <XR.h>
+
+#import <UIKit/UIKit.h>
+#import <ARKit/ARKit.h>
+#import <ARKit/ARConfiguration.h>
+
+extern void* GCALayerPtr;
+
+extern id<MTLDevice> MetalDevice;
+@interface SessionDelegate : NSObject <ARSessionDelegate>
+@property (readonly) id<MTLTexture> cameraTexture;
+@end
+
+@implementation SessionDelegate
+{
+    int width;
+    int height;
+    std::vector<xr::System::Session::Frame::View>* activeFrameViews;
+    CVMetalTextureCacheRef textureCache;
+    id<MTLTexture> _cameraTexture;
+}
+
+- (id)init:(std::vector<xr::System::Session::Frame::View>*)activeFrameViews metalContext:(id<MTLDevice>)graphicsContext
+{
+    self = [super init];
+    self->activeFrameViews = activeFrameViews;
+    
+    CVReturn err = CVMetalTextureCacheCreate(kCFAllocatorDefault, 0, graphicsContext, 0, &textureCache);
+    if (err)
+    {
+        // TODO: handle error
+    }
+    
+    return self;
+}
+
+- (id<MTLTexture>)updateCapturedTexture:(CVPixelBufferRef)pixelBuffer
+{
+    CVReturn ret = CVPixelBufferLockBaseAddress(pixelBuffer, kCVPixelBufferLock_ReadOnly);
+    if (ret != kCVReturnSuccess)
+    {
+        return {};
+    }
+    int planeIndex = 0;
+    width = (int) CVPixelBufferGetWidthOfPlane(pixelBuffer, planeIndex);
+    height = (int) CVPixelBufferGetHeightOfPlane(pixelBuffer, planeIndex);
+        
+    auto pixelFormat = MTLPixelFormatR8Unorm;
+    id<MTLTexture> mtlTexture;
+    CVMetalTextureRef texture;
+    auto status = CVMetalTextureCacheCreateTextureFromImage(kCFAllocatorDefault, textureCache, pixelBuffer, NULL, pixelFormat, width, height, planeIndex, &texture);
+    if (status == kCVReturnSuccess)
+    {
+        mtlTexture = CVMetalTextureGetTexture(texture);
+    }
+    
+    return mtlTexture;
+}
+
+- (UIInterfaceOrientation)orientation
+{
+    return [[UIApplication sharedApplication] statusBarOrientation];
+}
+
+- (CGSize)viewportSize
+{
+    UIInterfaceOrientation orientation = [self orientation];
+    CGSize viewport_size = CGSizeMake(width, height);
+    if (orientation == UIInterfaceOrientationPortrait || orientation == UIInterfaceOrientationPortraitUpsideDown)
+    {
+        if (width > height)
+        {
+            viewport_size = CGSizeMake(height, width);
+        }
+    }
+    else
+    {
+        if (width < height)
+        {
+            viewport_size = CGSizeMake(height, width);
+        }
+    }
+    return viewport_size;
+}
+
+- (void)session:(ARSession *)session didUpdateFrame:(ARFrame *)frame
+{
+    CVPixelBufferRef pixel_buffer = frame.capturedImage;
+    if (CVPixelBufferGetPlaneCount(pixel_buffer) < 2)
+    {
+        return;
+    }
+    
+    _cameraTexture = [self updateCapturedTexture:pixel_buffer];
```

Comment:
Does this need to be explicitly cleaned up, or does that happen automatically?

Reply:
It looks like that, yes. I'll check with the Xcode memory profiler to see if I missed something.